### PR TITLE
[RLlib] Fix issue 8158 ("PPO has no attribute 'workers'").

### PR DESCRIPTION
**Pull request by sven1977**
Fix for issue #8158.

#8158 

#8158 

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: rllib/agents/trainer.py [hunk 1]
```diff
@@ -1055,12 +1055,14 @@ def __getstate__(self):
         return state
 
     def __setstate__(self, state):
-        if "worker" in state:
+        if "worker" in state and hasattr(self, "workers") and \
+                isinstance(self.workers, WorkerSet):
```
Comment (ericl) [inline rllib/agents/trainer.py:1059]: Hmm do you know why PPO doesn't have the workers attribute? Shouldn't it always be present for that?

File: rllib/agents/trainer_template.py [hunk 2]
```diff
@@ -187,15 +187,15 @@ def _before_evaluate(self):
         def __getstate__(self):
             state = Trainer.__getstate__(self)
             state["trainer_state"] = self.state.copy()
-            if self.train_exec_impl:
+            if hasattr(self, "train_exec_impl") and self.train_exec_impl:
```
Comment (ericl) [inline rllib/agents/trainer_template.py:190]: Isn't this always set in init? How can it be missing?

File: rllib/agents/trainer.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/trainer.py:1059]: Very good question :) I have no idea at this point. I found lots of these checks, also for other props and therefore did not wonder much about it. But yeah, it seems like bad practice to not reliably have these in any instance of Trainer.

Comment (sven1977) [reply rllib/agents/trainer.py:1059]: I can dig further. May be a deeper issue that we should fix.

Comment (ericl) [reply rllib/agents/trainer.py:1059]: Sounds good. Maybe we are calling restore before the constructor is finished or something weird like that.

Comment (kanaadp) [reply rllib/agents/trainer.py:1059]: Any updates on this?

**Pull request closed without merging**